src/periodic.jl

```diff
@@ -0,0 +1,93 @@
+"""
+    Periodic{T}(f, period::T, phase_shift::T) <: Transform
+
+Applies a periodic function `f` with provided `period` and `phase_shift` to the data.
+
+!!! note
+    For `TimeType` data, the result will change depending on the type of `period` given,
+    even if the same amount of time is described. Example: `Week(1)` vs `Second(Week(1))`;
+    the former starts the period on the most recent Monday, while the latter starts the
+    period on the most recent multiple of 604800 seconds since time 0.
+
+# Fields
+* `f::Union{typeof(cos), typeof(sin)}`: the periodic function
+* `period::Union{Real, Period}`: the function period. Must be strictly positive.
+* `phase_shift::Union{Real, Period}`: adjusts the phase of the periodic function, measured
+    in the same units as the input. Increasing the value translates the function to the
+    right, toward higher/later input values.
+"""
+struct Periodic{T} <: Transform where T <: Union{Real, Period}
+    f::Union{typeof(cos), typeof(sin)}
+    period::T
+    phase_shift::T
+
+    function Periodic(f, period::T, phase_shift::T) where T
+        period <= zero(T) && throw(ArgumentError("period must be strictly positive."))
+        return new{T}(f, period, phase_shift)
+    end
+end
+
+"""
+    Periodic(f, period) -> Periodic
+
+A constructor for [`Periodic`](@ref).
+Returns a `Periodic` transform with zero phase shift.
+"""
+Periodic(f, period::T) where T = Periodic(f, period, zero(T))
+
+function _apply!(x::AbstractArray{T}, P::Periodic; kwargs...) where T <: Real
+    x[:] = P.f.(2π .* (x .- P.phase_shift) / P.period)
+    return x
+end
+
+function apply(x::AbstractArray{T}, P::Periodic{U}; kwargs...) where {T <: TimeType, U <: Period}
+    return map(xi -> _periodic(P.f, xi, P.period, P.phase_shift), x)
+end
+
+"""
+    Transforms.apply(table, ::Periodic{T}; cols=nothing) where T <: Period -> Array
+
+Applies [`Periodic`](@ref) to each of the specified columns in `table`.
+If no `cols` are specified, then [`Periodic`](@ref) is applied to all columns.
+Returns an array containing each transformed column, in the same order as `cols`.
+"""
+function apply(table, P::Periodic{T}; cols=nothing) where T <: Period
+    Tables.istable(table) || throw(MethodError(apply, (table, P)))
+
+    columntable = Tables.columns(table)
+    cnames = cols === nothing ? propertynames(columntable) : cols
+    return [apply(getproperty(columntable, cname), P) for cname in cnames]
+end
+
+"""
+    _periodic(f, instant, period, phase_shift=Day(0))
+
+Computes the value of periodic function `f` at the given instant in time.
+
+# Arguments
+* `f`: the periodic function
+* `period`: the function period
+* `phase_shift`: adjusts the phase of the periodic function. Increasing the value translates
+    the function to the right, toward higher/later input values.
+"""
+function _periodic(f, instant, period, phase_shift=Day(0))
```

Comment:
I've only just noticed that we use this for `TimeTypes` but not for reals... Do we know why this is? might be worth explaning in the docstring

Reply:
I think it stems from the fact that the `instant`s we tend to pass in are far from the "origin" of time (`DateTime(0)`) and so `(x .- t) / period` is a huge number and calling `sin` / `cos` on this will inevitably introduce rounding errors even if `sin(0) - sin(2\pi * N) = 0` computationally this diverges as N grows large.

So to compensate we define an origin closer to the `instant` which, by virtue of periodicity, is equivalent to `DateTime(0)`.

Reply:
I will leave handling this similarly for `Real` to another time, making an issue. We generally expect the `TimeType` inputs to be much further from the origin, but there's nothing stopping large inputs for `Real` either.